comment/api/serializers.py

```diff
@@ -99,25 +99,21 @@ class Meta:
         model = Comment
         fields = ('id', 'user', 'email', 'content', 'parent', 'posted', 'edited', 'reply_count', 'replies', 'urlhash')
 
-    def __init__(self, *args, **kwargs):
-        user = kwargs['context']['request'].user
-        self.email_service = None
-        if user.is_authenticated or not settings.COMMENT_ALLOW_ANONYMOUS:
-            del self.fields['email']
-
-        super().__init__(*args, **kwargs)
-
     @staticmethod
-    def validate_email(value):
-        if not value:
-            raise serializers.ValidationError(EmailError.EMAIL_MISSING, code='required')
-        return value.strip().lower()
+    def validate_email(email):
+        if not email:
+            raise serializers.ValidationError(
+                detail={'email': [EmailError.EMAIL_REQUIRED_FOR_ANONYMOUS]}, code='required'
+            )
+        return email.strip().lower()
 
     def create(self, validated_data):
         request = self.context['request']
         user = get_user_for_request(request)
         content = validated_data.get('content')
         email = validated_data.get('email')
```

Comment:
`validate_email` is already called here. You can try to enter a email like `A@A.COM` with an anonymous user and see the corresponding value in the database. If the value saved inside the database is `a@a.com`, you won't need to call `validate_email` manually.

Reply:
You need to call it here otherwise, missing email won't be validated.

Reply:
This is called by default as state in the documentation about [field level validation](https://www.django-rest-framework.org/api-guide/serializers/#field-level-validation)

```sh
> git branch --show-current                                                                                                                           
develop                                                                                                                                               
git❨ develop ❩                                                                                                                                        
curious@tinker:pts/16-> /home » curious » Dev » opensourceDev » Comment (0)                                                                           
> djshell                                                                                                                                             
Python 3.8.5 (default, Jan 27 2021, 15:41:15)                                                                                                         
[GCC 9.3.0] on linux 
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> from comment.api.serializers import CommentCreateSerializer as CS
>>> from django.test import RequestFactory as RF
>>> from post.models import Post
>>>
>>> post = Post.objects.last()
>>>
>>> from django.contrib.auth.models import AnonymousUser
>>>
>>> request = RF().get('/')
>>> request.user = AnonymousUser()
>>> context = {'model_obj': post, 'request': request}
>>> data = {'email': '', 'content': 'test'}
>>>
>>> from comment.conf import settings
>>> settings.COMMENT_ALLOW_ANONYMOUS = True
>>>
>>> serializer = CS(context=context, data=data)
>>> serializer.is_valid()
False
>>> serializer.data
{'email': '', 'content': 'test'}
>>> serializer.errors
{'email': [ErrorDetail(string='Email is required for posting anonymous comments.', code='required')]}
```

Reply:
when we access `validated_data`, the methods for `validate_<field>` have already been called by  restframeworkk's `serializers.Serializer` class.

Reply:
Well, it seems that `serializer.is_valid()` is not called. Try this
```bash
curl -X POST -H "Content-Type: application/json" -d '{"content": "test"}' "$BASE_URL/api/comments/create/?app_name=post&model_name=post&model_id=1"
```

Reply:
yes, this seems to be an unintended behavior from our side, we can probably merge this. I will try to investigate more about this later.

Reply:
We have discussed this previously. [REST docs](https://www.django-rest-framework.org/api-guide/serializers/#field-level-validation)
> Note: If your <field_name> is declared on your serializer with the parameter required=False then this validation step will not take place if the field is not included.

In our case, the email field is not required by default but it's included.

Reply:
the validation won't take place if the field is not included, but it is included for unauthenticated users.